encoding/cbf/cbf_codec/encode_value.go

```diff
@@ -0,0 +1,394 @@
+		if err != nil {
+			return
+		}
+		return common_codec.EncodeNumber(&e.w, uint16(v))
+	case cadence.Word32:
+		err = e.EncodeValueIdentifier(EncodedValueWord32)
+		if err != nil {
+			return
+		}
+		return common_codec.EncodeNumber(&e.w, uint32(v))
+	case cadence.Word64:
+		err = e.EncodeValueIdentifier(EncodedValueWord64)
+		if err != nil {
+			return
+		}
+		return common_codec.EncodeNumber(&e.w, uint64(v))
+	case cadence.Fix64:
+		err = e.EncodeValueIdentifier(EncodedValueFix64)
+		if err != nil {
+			return
+		}
+		return common_codec.EncodeNumber(&e.w, int64(v))
+	case cadence.UFix64:
+		err = e.EncodeValueIdentifier(EncodedValueUFix64)
+		if err != nil {
+			return
+		}
+		return common_codec.EncodeNumber(&e.w, uint64(v))
+	case cadence.Array:
+		switch arrayType := v.ArrayType.(type) {
+		case cadence.VariableSizedArrayType:
+			err = e.EncodeValueIdentifier(EncodedValueVariableArray)
+			if err != nil {
+				return
+			}
+			err = e.EncodeVariableArrayType(arrayType)
+		case cadence.ConstantSizedArrayType:
+			err = e.EncodeValueIdentifier(EncodedValueConstantArray)
+			if err != nil {
+				return
+			}
+			err = e.EncodeConstantArrayType(arrayType)
+		case nil:
+			err = e.EncodeValueIdentifier(EncodedValueUntypedArray)
+		default:
+			err = common_codec.CodecError(fmt.Sprintf("unknown array type: %s", arrayType))
+		}
+		if err != nil {
+			return
+		}
+		return e.EncodeArray(v)
+	case cadence.Dictionary:
+		err = e.EncodeValueIdentifier(EncodedValueDictionary)
+		if err != nil {
+			return
+		}
+		return e.EncodeDictionary(v)
+	case cadence.Struct:
+		err = e.EncodeValueIdentifier(EncodedValueStruct)
+		if err != nil {
+			return
+		}
+		return e.EncodeStruct(v)
+	case cadence.Resource:
+		err = e.EncodeValueIdentifier(EncodedValueResource)
+		if err != nil {
+			return
+		}
+		return e.EncodeResource(v)
+	case cadence.Event:
+		err = e.EncodeValueIdentifier(EncodedValueEvent)
+		if err != nil {
+			return
+		}
+		return e.EncodeEvent(v)
+	case cadence.Contract:
+		err = e.EncodeValueIdentifier(EncodedValueContract)
+		if err != nil {
+			return
+		}
+		return e.EncodeContract(v)
+	case cadence.Link:
+		err = e.EncodeValueIdentifier(EncodedValueLink)
+		if err != nil {
+			return
+		}
+		return e.EncodeLink(v)
+	case cadence.Path:
+		err = e.EncodeValueIdentifier(EncodedValuePath)
+		if err != nil {
+			return
+		}
+		return e.EncodePath(v)
+	case cadence.Capability:
+		err = e.EncodeValueIdentifier(EncodedValueCapability)
+		if err != nil {
+			return
+		}
+		return e.EncodeCapability(v)
+	case cadence.Enum:
+		err = e.EncodeValueIdentifier(EncodedValueEnum)
+		if err != nil {
+			return
+		}
+		return e.EncodeEnum(v)
+	}
+
+	return common_codec.CodecError(fmt.Sprintf("unexpected value: %s (type=%s)", value, value.Type()))
+}
+
+func (e *Encoder) EncodeValueIdentifier(id EncodedValue) (err error) {
+	return e.writeByte(byte(id))
+}
+
+func (e *Encoder) EncodeOptional(value cadence.Optional) (err error) {
```

Comment:
will cadence.Optional always be represented as `nil | some value`? might be worth adding a method on the Optional type for checking if it's inhabited, in case the internal implementation changes